test/integration/consul-container/test/envoy_extensions/wasm_test.go

```diff
@@ -0,0 +1,626 @@
+// serve the compiled wasm extension. The static proxy is configured to apply the wasm filter
+// and pointed at the remote file server. It should download the file and add it to the envoy filters.
+func TestWASMRemote(t *testing.T) {
+	t.Parallel()
+
+	// build all the file paths we will need for the test
+	cwd, err := os.Getwd()
+	require.NoError(t, err, "could not get current working directory")
+	hostWASMDir := fmt.Sprintf("%s/testdata/wasm_test_files", cwd)
+
+	buildWASM(t, hostWASMDir)
+
+	cluster, _, _ := topology.NewCluster(t, &topology.ClusterConfig{
+		NumServers:                1,
+		NumClients:                1,
+		ApplyDefaultProxySettings: true,
+		BuildOpts: &libcluster.BuildOptions{
+			Datacenter:             "dc1",
+			InjectAutoEncryption:   true,
+			InjectGossipEncryption: true,
+		},
+	})
+
+	clientService, staticProxy := createServices(t, cluster)
+	_, port := clientService.GetAddr()
+	_, adminPort := clientService.GetAdminAddr()
+
+	libassert.AssertUpstreamEndpointStatus(t, adminPort, "static-server.default", "HEALTHY", 1)
+	libassert.GetEnvoyListenerTCPFilters(t, adminPort)
+
+	libassert.AssertContainerState(t, clientService, "running")
+	libassert.AssertFortioName(t, fmt.Sprintf("http://localhost:%d", port), "static-server", "")
+
+	// Check header not present
+	c1 := cleanhttp.DefaultClient()
+	res, err := c1.Get(fmt.Sprintf("http://localhost:%d", port))
+	require.NoError(t, err)
+
+	// check that header DOES NOT exist before wasm applied
+	if value := res.Header.Get(http.CanonicalHeaderKey("x-test")); value != "" {
+		t.Fatal("unexpected test header present before WASM applied")
+	}
+
+	// Create Nginx file server
+	uri, nginxService, nginxProxy := createNginxFileServer(t, cluster,
+		// conf file
+		testcontainers.ContainerFile{
+			HostFilePath:      fmt.Sprintf("%s/nginx.conf", hostWASMDir),
+			ContainerFilePath: "/etc/nginx/conf.d/wasm.conf",
+			FileMode:          777,
+		},
+		// extra files loaded after startup
+		testcontainers.ContainerFile{
+			HostFilePath:      fmt.Sprintf("%s/wasm_add_header.go.wasm", hostWASMDir),
+			ContainerFilePath: "/usr/share/nginx/html/wasm_add_header.wasm",
+			FileMode:          777,
+		})
+
+	defer nginxService.Terminate()
+	defer nginxProxy.Terminate()
+
+	// wire up the wasm filter
+	node := cluster.Agents[0]
+	client := node.GetClient()
+
+	agentService, _, err := client.Agent().Service(libservice.StaticServerServiceName, nil)
+	require.NoError(t, err)
+
+	agentService.Connect = &api.AgentServiceConnect{
+		SidecarService: &api.AgentServiceRegistration{
+			Kind: api.ServiceKindConnectProxy,
+			Proxy: &api.AgentServiceConnectProxyConfig{
+				Upstreams: []api.Upstream{
+					{
+						DestinationName:  "nginx-fileserver",
+						DestinationPeer:  "",
+						LocalBindAddress: "0.0.0.0",
+						LocalBindPort:    9595,
+					},
+				},
+			},
+		},
+	}
+
+	// Upsert the service registration to add the upstreams
+	// TODO(@landa) there has to be a better way to add upstreams
```

Comment:
Unfortunately there isn't :joy: 

```suggestion
	// Upsert the service registration to add the nginx-fileserver as an
	// upstream of the static-server so that it can retrieve the WASM plugin
```